### PR TITLE
Safer container names

### LINKED ISSUE #5680: [Security] Docker socket exported writable; better use docker socket proxy
##### Bug report summary
Current documentation for running netdata in a container recommends to export the docker socket volume:

>   -v /var/run/docker.sock:/var/run/docker.sock:ro \

First, the `:ro` does not what people expect: you can still access it and write to the socket. You can just not change the file. So I recommend to either remove it to not lull users to safety, or at least document clearly what it does (which is, nothing useful).
Second, exporting the volume into a container with outside access is a very, very bad idea, see:

- [Its a good thing SELinux blocks access to the docker socket.
](https://danwalsh.livejournal.com/74095.html)
- [Don't expose the Docker socket (not even to a container)](https://www.lvh.io/posts/dont-expose-the-docker-socket-not-even-to-a-container.html)

tl;dr: With this you enable direct root access to the host for anyone taking over anything in your container. This is less secure running netdata outside the container framework, as a normal application user.

Instead, please add a docker socket proxy which is not reachable from the outside, but exports the docker capabilities in a controllable and safe way - and read only. See for example:

- [Docker Socket Proxy](https://github.com/Tecnativa/docker-socket-proxy/blob/master/README.md)
- [[Howto] Launch traefik as a docker container in a secure way](https://liquidat.wordpress.com/2018/12/12/howto-launch-traefik-as-a-docker-container-in-a-secure-way/) was an example how to do it with other tools which need docker socket access

##### OS / Environment
Docker

##### Steps To Reproduce

- Start netdata container as documented with the socket exported.
- Once inside the container, start arbitrary containers taking over the host.

##### Expected behavior

- Access to docker socket should be done in a secure way.
- The documentation should not lull to safety with options which do not do what people think.

### LINKED ISSUE #6293: netdata/installer/docker: Question regarding docker container name resolution config
##### Question summary
Hello @ArjenR 
Following up our discussion [here](https://github.com/netdata/netdata/commit/7ee5dc37ae6bca9945db518b2b0cc7d69ce36ae0#r33952591), i am creating this issue to continue the discussion.
I 'll post my next comment below.

##### OS / Environment
Any

##### Component Name
netdata/installer/docker

### PR DESCRIPTION
**Pull request by ViViDboarder**
##### Summary

The current method of getting Docker container names is risky and non-standard. This is documented in #5680 and #6293. This new diff will allow a safer method of querying as well as allow the user of a proxy between the docker socket (as described on #5680).

There is one issue I encountered when testing. For some reason, when I run the application configured to access `/var/run/docker.sock`, I'm seeing `curl: (7) Couldn't connect to server`. However, if I run the plugin directly via `docker exec`, it works perfectly. Eg: `/usr/libexec/netdata/plugins.d/cgroup-name.sh docker_a1c9bbf558a66841ba5b9b465ee3d6e9f6369cab6423e2fd8fb2a931e0adc407`

I'm not sure if there is some other complication due to how the script is being executed, but any pointers on where to go to resolve that would be appreciated.

I did confirm that it's not a bug introduced by my diff as I get the same issue using `nc`.

##### Additional Information

This branch actually contains two diffs, one to make testing easier and the other to actually implement the change. I'm happy to revert the testing one or force remove it, but thought it might be handy for others.
Fixes #5680 
Fixes #6293

### CODE REVIEW

**Comment by paulkatsoulakis** [comment]
Hello @ViViDboarder, thanks a lot for your PR! I will look into it and let you know.

On a first thought, i am suspecting you are facing the same problems i tried to mitigate with the docker socket, but i will let you know once i do the tests and see what happening

**Comment by paulkatsoulakis** [comment]
Hello @ViViDboarder, thanks a lot for your PR! I will look into it and let you know.

On a first draft look, i think i see a couple of problems.
I will take the branch offline and work it out before i give you some further feedback

File: packaging/docker/run.sh [hunk 1]
```diff
@@ -34,18 +34,6 @@ create_group_and_assign_to_user() {
 	fi
 }
 
-DOCKER_USR="netdata"
```
Comment (paulkatsoulakis) [inline packaging/docker/run.sh]: Do not remove this variable, it breaks netdata.
We use it elsewhere too

File: packaging/docker/run.sh [hunk 2]
```diff
@@ -34,18 +34,6 @@ create_group_and_assign_to_user() {
...
-DOCKER_SOCKET="/var/run/docker.sock"
-DOCKER_GROUP="docker"
-
-if [ -S "${DOCKER_SOCKET}" ] && [ -n "${PGID}" ]; then
-	GRP=$(create_group_and_assign_to_user "${DOCKER_GROUP}" "${PGID}" "${DOCKER_USR}")
```
Comment (paulkatsoulakis) [inline packaging/docker/run.sh]: Since you removed `create_group_and_assign_to_user`, kindly remove the declaration of that function above, we do not use it elsewhere

File: packaging/docker/build-test.sh [hunk 3]
```diff
@@ -46,27 +46,29 @@ do
 	esac
 done
 
-if [ -n "${REPOSITORY}" ] && [ -n "${VERSION}" ] && [ -n "${DOCKER_USERNAME}" ] && [ -n "${DOCKER_PWD}" ] ; then
```
Comment (paulkatsoulakis) [inline packaging/docker/build-test.sh]: This is intentional, so that we can choose the version we want to run.
Unless there is a very specific reason for it, kindly re-instate this (along with the respective references on the messages printed)

File: collectors/cgroups.plugin/cgroup-name.sh.in [hunk 4]
```diff
@@ -53,18 +53,26 @@ function docker_get_name_classic() {
...
 		warning "Can't find jq command line tool. jq is required for netdata to retrieve docker container name using ${DOCKER_HOST} API, falling back to docker ps"
 		return 1
 	fi
-
-	info "Running API command: /containers/${id}/json"
-	JSON=$(echo -e "GET /containers/${id}/json HTTP/1.0\\r\\n" | nc -U "${DOCKER_HOST}" | grep '^{.*')
+    if [ -S "${DOCKER_HOST}" ]; then
+        info "Running API command: curl --unix-socket ${DOCKER_HOST} http://localhost${path}"
+        JSON=$(curl --unix-socket ${DOCKER_HOST} http://localhost${path})
+    elif [ "${DOCKER_HOST}" == "/var/run/docker.sock" ]; then
```
Comment (paulkatsoulakis) [inline collectors/cgroups.plugin/cgroup-name.sh.in]: remove this else-if.
if DOCKER_HOST is not a socket, we just go for the API approach

File: collectors/cgroups.plugin/cgroup-name.sh.in [hunk 5]
```diff
@@ -53,18 +53,26 @@ function docker_get_name_classic() {
...
+        warning "Docker socket was not found at ${DOCKER_HOST}"
+        return 1
+    else
+        info "Running API command: curl ${DOCKER_HOST}${path}"
```
Comment (paulkatsoulakis) [inline collectors/cgroups.plugin/cgroup-name.sh.in]: In here, before we run the API, you can do the check:

if [ -f ${DOCKER_SOCKET} ]; then
      echo "Found ${DOCKER_SOCKET} as a file, i don't know what to do"
      return 1
else
      API CALL CODE HERE
fi

File: collectors/cgroups.plugin/cgroup-name.sh.in [hunk 6]
```diff
@@ -53,18 +53,26 @@ function docker_get_name_classic() {
...
 	if ! command -v jq >/dev/null 2>&1; then
 		warning "Can't find jq command line tool. jq is required for netdata to retrieve docker container name using ${DOCKER_HOST} API, falling back to docker ps"
 		return 1
 	fi
-
-	info "Running API command: /containers/${id}/json"
-	JSON=$(echo -e "GET /containers/${id}/json HTTP/1.0\\r\\n" | nc -U "${DOCKER_HOST}" | grep '^{.*')
+    if [ -S "${DOCKER_HOST}" ]; then
+        info "Running API command: curl --unix-socket ${DOCKER_HOST} http://localhost${path}"
+        JSON=$(curl --unix-socket ${DOCKER_HOST} http://localhost${path})
```
Comment (paulkatsoulakis) [inline collectors/cgroups.plugin/cgroup-name.sh.in]: Add quotes around ${DOCKER_HOST} (This is caught from travis)

File: collectors/cgroups.plugin/cgroup-name.sh.in [hunk 7]
```diff
@@ -53,18 +53,26 @@ function docker_get_name_classic() {
...
+    elif [ "${DOCKER_HOST}" == "/var/run/docker.sock" ]; then
+        warning "Docker socket was not found at ${DOCKER_HOST}"
+        return 1
+    else
+        info "Running API command: curl ${DOCKER_HOST}${path}"
+        JSON=$(curl ${DOCKER_HOST}${path})
```
Comment (paulkatsoulakis) [inline collectors/cgroups.plugin/cgroup-name.sh.in]: Same, add quotes around "${DOCKER_HOST}${path}"

File: packaging/docker/run.sh (hunk 1 above)
Comment (ViViDboarder) [reply packaging/docker/run.sh]: Got it. Didn't realize this is used to tell NetData to run as the `netdata` user. I'm pretty sure this is actually why I'm seeing the issues with accessing the socket from within NetData, but not when I `exec`, since `exec` drops me into `root`.

Rather than chowning the Docker socket, I'm can add some instructions on how to pass in an existing `GID` that has access. The reason I believe this is safer is that access to the socket is essentially root access to the entire system. Any change of ownership is significant and should be done very deliberately by the user.

 Alternately, a variable could be used to tell NetData to run as `root` instead. There is no difference in safety since, as described earlier, access to the socket allows root escalation already.

File: packaging/docker/build-test.sh (hunk 3 above)
Comment (ViViDboarder) [reply packaging/docker/build-test.sh]: It is still possible to provide all the variables. I believe all existing functionality is still preserved. This change only makes it all optional so that when doing local development the script will not fail when it attempts to log into Docker.

It doesn't look like this script really adds too much and potentially a simple `docker build` command could suffice for local development. I can revert this whole file if you're prefer to keep all these arguments as mandatory.

File: collectors/cgroups.plugin/cgroup-name.sh.in (hunk 4 above)
Comment (ViViDboarder) [reply collectors/cgroups.plugin/cgroup-name.sh.in]: The reason I have this extra `elif` is because the default value is `/var/run/docker.sock`. If a user provides no env variables but also does not mount the socket, they will fall into this case and see some message.

I could make it say something else or direct them to the documentation (I'll rewrite and publish in a new diff) that explains how to provide access to Docker.

File: collectors/cgroups.plugin/cgroup-name.sh.in (hunk 5 above)
Comment (ViViDboarder) [reply collectors/cgroups.plugin/cgroup-name.sh.in]: This is definitely another case we could catch, but I feel that it's probably more likely to be an accidental directory rather than a file. If someone provides a Docker volume to a path that does not exist, I believe it ends up getting created as a directory.

You think I should add that?

**Comment by ViViDboarder** [comment]
Travis failure doesn't appear to be related: `ABORTED  Cannot download https://storage.googleapis.com/netdata-nightlies/netdata-latest.tar.gz`

The branch seems to be working as intended and I verified the instructions provided by the new documentation.

I'm happy to update the error cases and revert the build script changes, if desired.

File: packaging/docker/run.sh (hunk 1 above)
Comment (paulkatsoulakis) [reply packaging/docker/run.sh]: Hi, Please re-instate this variable. Netdata is designed to **not** run as root.
We control the level of privileged access is required on a per module basis.

Reference:
https://docs.netdata.cloud/docs/netdata-security/?origin_team=TCC5DL092#your-systems-are-safe-with-netdata

**Comment by ViViDboarder** [comment]
@paulkatsoulakis the variable is there. It's just been moved into the `Dockerfile` since that's where the user gets created as well. That way creation and use are referenced from the same variable.

**Comment by ViViDboarder** [comment]
Also, in regards to running as non-root for security, that is voided as soon as you give the `netdata` user access to the Docker socket. From that point forward, the `netdata` user has root access to the machine through the Docker daemon. (eg. `docker run -v /:/host busybox`)

The motivation for this diff is to make that clear and provide an alternative way to access the socket.

Another way that I could implement this to provide access as a non-root user (though to be clear, they'd still have root access, just not as UID=0) without changing ownership of the socket itself, would be to generate a user within the Docker entrypoint with the same UID and GUID of a user that already has access. Though I believe I tried this and ran into some issues with the socket being mounted.

I can do some more research on alternatives.

File: packaging/docker/README.md [hunk 8]
```diff
@@ -49,33 +49,55 @@ services:
...
-   /var/run/docker.sock:/var/run/docker.sock:ro
-   ```
-
-   change to
-   ```
-   /var/run/docker.sock:/var/run/docker.sock:rw
-   ```
+There are a few options for resolving container names within netdata. Some methods of doing so will allow root access to your machine from within the container. Please read the following carefully.
+
+#### Docker Socket Proxy (Safest Option)
```
Comment (paulkatsoulakis) [inline packaging/docker/README.md]: Hello, you will have to open a github issue with the proposed proxy, so that we evaluate and test before we recommend it as part of netdata's solution.  We can't put the effort through a PR, it should rather be a proper investigation for this.

Kindly revert the recommendation so that we can merge the fix, thanks!

Comment (ViViDboarder) [reply packaging/docker/README.md]: Remove references to the image that I tested with. A separate PR makes sense.

**Comment by paulkatsoulakis** [comment]
Hello @ViViDboarder, you still need to sign the CLA in order to proceed.

Thanks again for the PR!

**Comment by ViViDboarder** [comment]
@joelhans I just reverted the diff that included the example. @paulkatsoulakis not sure how much time it takes to review the image that I'm recommending, but Joel has a point that it would be helpful for users to implement. The image is the top one recommended when Ducking "docker socket proxy", so I imagine a savvy user will end up with the same image.

**Comment by ViViDboarder** [comment]
Just updated to include instructions described in #6508 as well.

**Comment by paulkatsoulakis** [comment]
> @joelhans I just reverted the diff that included the example. @paulkatsoulakis not sure how much time it takes to review the image that I'm recommending, but Joel has a point that it would be helpful for users to implement. The image is the top one recommended when Ducking "docker socket proxy", so I imagine a savvy user will end up with the same image.

I think we can put a note that we are working on providing a recommendation on the proxy option.
@joelhans you have a very good point, the thing is we need to evaluate whatever solution we will recommend and that is why i preferred we created a separate issue to investigate the option @ViViDboarder has in mind, before we pop it on the documentation

**Comment by ViViDboarder** [comment]
@paulkatsoulakis great. I actually do refer to HAProxy in the documentation already, but caught a typo. So I've updated that.

Let me know if there is anything you'd like me to update or clarify.

File: packaging/docker/README.md [hunk 9]
```diff
@@ -49,33 +49,51 @@ services:
     volumes:
       - /proc:/host/proc:ro
       - /sys:/host/sys:ro
-      - /var/run/docker.sock:/var/run/docker.sock:ro
```
Comment (paulkatsoulakis) [inline packaging/docker/README.md]: There is one more place above that refers to the socket.
Sorry i missed it on the first sweep

File: packaging/docker/README.md [hunk 10]
```diff
@@ -49,33 +49,51 @@ services:
...
-
-   change to
-   ```
-   /var/run/docker.sock:/var/run/docker.sock:rw
-   ```
+There are a few options for resolving container names within netdata. Some methods of doing so will allow root access to your machine from within the container. Please read the following carefully.
+
+#### Docker Socket Proxy (Safest Option)
+
+Deploy a Docker socket proxy that accepts will filter out requests using something like HAProxy so that it restricts connections to read-only access to the CONTAINERS endpoint.
```
Comment (paulkatsoulakis) [inline packaging/docker/README.md]: `Deploy a Docker socket proxy that accepts will filter out requests` -> `Deploy a Docker socket proxy that accepts and filter out requests` unless you meant something else.

Also change HAProxy to [HAProxy](https://docs.netdata.cloud/docs/running-behind-haproxy/) so that it links to our documentation

Comment (ViViDboarder) [reply packaging/docker/README.md]: I made this change, but I do worry it might be a little confusing as that documentation doesn't explain how to use HAProxy for this purpose. Though that's a good place for the documentation to be updated to explain it!

File: packaging/docker/Dockerfile [hunk 11]
```diff
@@ -58,9 +58,11 @@ COPY --from=builder /app /
 # Configure system
 ARG NETDATA_UID=201
 ARG NETDATA_GID=201
+ENV DOCKER_GRP netdata
+ENV DOCKER_USR netdata
 RUN \
     # provide judy installation to base image
-    apk add make alpine-sdk && \
+    apk add make alpine-sdk shadow && \
```
Comment (paulkatsoulakis) [inline packaging/docker/Dockerfile]: I was wondering, why do we need shadow package?

Comment (ViViDboarder) [reply packaging/docker/Dockerfile]: That's to add the `usermod` command for adding the `netdata` user to the `docker` group on run.

https://github.com/netdata/netdata/pull/6441/files#diff-8ab2a2dd9f9a25a877e05f48b46a7ee3R20
